### ISSUE TITLE
How to prevent re-rendering of component when selectorFamily state remains unchanged?

### ISSUE THREAD

**Issue by ErlendEllingsen** [facebookexperimental/Recoil#522]
Hi, 

Really love Recoil, and have started using it in a project quite extensively. I clearly see the potential of this state mgmt library. 
I have setup an app with multiple atoms and selectors, and it all works very smooth, however, I have one scenario that I have yet to solve in a clever manner, and I have not been able to find more info about it on your docs.

I have a table with multiple rows. Each row is a React.memo sub-component that uses Recoil hooks. The Recoil part of the component looks as follows (excuse my variable namings): 

```tsx
const setMarkInvalidModalState = useSetRecoilState(markInvalidModalState)
const setMarkInvalidSelectTrade = useSetRecoilState(markInvalidSelectTradeState)

const setOpenTradeModal = useSetRecoilState(openTradeModalState)
const setOpenTradeSelectTrade = useSetRecoilState(openTradeSelectTradeState)

const markForTrade = useRecoilValue(getMarkForTrade(compositeKey));
```

The only place i subscribe for updates and re-renderings is useRecoilValue of getMarkForTrade. `getMarkForTrade` is a `selectorFamily`, that looks as follows:

```ts
const getMarkForTrade = selectorFamily({
  key: 'MarkForTradeState',
  get: (compositeKey: string) => ({get}) => {
    const allMarks = get(getTradeMarks);

    const mark = allMarks[compositeKey];
    return ( mark !== undefined ? mark : null) as TradeMark | null;
  },
});
```

Now, whenever the underlying selector `getTradeMarks` (called in `getMarkForTrade`) is updated, all components subscribed to `getMarkForTrade` is also getting re-rendered. Even though the output / state of `getMarkForTrade` has remained unchanged for the input. 

Basically all rows in the table are re-rendered every time, although 90% of the rows have the same output state from `getMarkForTrade`.

How can I prevent this? I have tested multiple times that this is the selector that triggers the re-render.

**Comment by manyopensource** [comment]
Could you provide the minimal reproducible sample?

**Comment by wuzzeb** [comment]
I've been using this in the same situation, a table where only the row that was changed gets re-rendered.

```typescript
export function useMemoizedRecoilValue<T>(recoilValue: RecoilValueReadOnly<T>): T {
  const [, forceRender] = React.useReducer((s: number) => s + 1, 0);
  const lastVal = React.useRef<T>();
  const lastRecoilValue = React.useRef<RecoilValueReadOnly<T>>();
  const init = useRecoilCallback(
    ({ snapshot }) => () => {
      lastVal.current = snapshot.getLoadable(recoilValue).contents as T;
      lastRecoilValue.current = recoilValue;
    },
    [recoilValue]
  );
  useRecoilTransactionObserver_UNSTABLE(({ snapshot }) => {
    const cur = snapshot.getLoadable(recoilValue).contents as T;
    if (cur !== lastVal.current || lastRecoilValue.current !== recoilValue) {
      lastVal.current = cur;
      forceRender();
    }
  });

  if (lastRecoilValue.current !== recoilValue) {
    init();
  }

  return lastVal.current as T;
}
```

Doesn't work with promises and just uses !== to compare, but both could be supported by simple changes to the code.

In my opinion, the best built-in solution would be that the selector is passed the old value to the get function and the selector can implement any checks or whatnot itself.   You can see in #314 that it is hard to implement a generic method of suppressing updates so my idea is:

- selector is passed the old value
- if the selector returns the exact same javascript object passed to old, downstream updates are suppressed.  This is just like the recent change to atoms which suppressed updates with the exact same object is set.

This way selectors that  don't care just ignore the old input and the few selectors which need to suppress downstream updates can implement checks at the beginning of the function.

**Comment by drarmstr** [comment]
Duplicate of #314

**Comment by fasiha** [comment]
Here's my minimal example: I started with these:

```console
$ npx create-react-app my-app
$ cd my-app
$ npm install recoil
$ npm start
```

Then I modified App.js per the following commit: https://github.com/fasiha/recoil-issue-522/commit/623c65b51cf43ab487c9eeba7447e8e8a7a54efc#diff-14b1e33d5bf5649597cdc0e4f684dadd (profound apologies, I don't know how to JSX, I just use `createElement` aliased to `ce` here, hopefully that doesn't distract too much)

The build is available at http://fasiha.github.io/recoil-issue-522

Here's how you can see that, even though only a single member of the selector family should be updating (as I click the "Add" buttom), *all* rows rerender:

![output](https://user-images.githubusercontent.com/37649/90535545-01f64380-e149-11ea-88e0-40d42f6d47aa.gif)

**Comment by drarmstr** [comment]
@wuzzeb - See optimization in #749

**Comment by salvoravida** [comment]
maybe related https://github.com/facebookexperimental/Recoil/pull/825
